csds-client/client/client.go

```diff
@@ -0,0 +1,255 @@
+		configFile:  *configFilePtr,
+		monitorFreq: *monitorFreqPtr,
+	}
+
+	return f
+}
+
+// parseNodeMatcher parses the csds request yaml to nodematcher
+func (c *Client) parseNodeMatcher() error {
+	if c.info.requestFile == "" && c.info.requestYaml == "" {
+		return fmt.Errorf("missing request yaml")
+	}
+
+	var nodematchers []*envoy_type_matcher.NodeMatcher
+	if err := parseYaml(c.info.requestFile, c.info.requestYaml, &nodematchers); err != nil {
+		return err
+	}
+
+	c.nm = nodematchers
+
+	// check if required fields exist in nodematcher
+	switch c.info.platform {
+	case "gcp":
+		switch c.info.uri {
+		case "trafficdirector.googleapis.com:443":
+			keys := []string{"TRAFFICDIRECTOR_GCP_PROJECT_NUMBER", "TRAFFICDIRECTOR_NETWORK_NAME"}
+			for _, key := range keys {
+				if value := getValueByKeyFromNodeMatcher(c.nm, key); value == "" {
+					return fmt.Errorf("Missing field %v in NodeMatcher", key)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// connWithAuth connects to uri with authentication
+func (c *Client) connWithAuth() error {
+	var scope string
+	switch c.info.authnMode {
+	case "jwt":
+		if c.info.jwt == "" {
+			return fmt.Errorf("missing jwt file")
+		}
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewServiceAccountFromFile(c.info.jwt, scope)
+			if err != nil {
+				return err
+			}
+
+			c.cc, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return nil
+		}
+	case "auto":
+		switch c.info.platform {
+		case "gcp":
+			scope = "https://www.googleapis.com/auth/cloud-platform"
+			pool, err := x509.SystemCertPool()
+			creds := credentials.NewClientTLSFromCert(pool, "")
+			perRPC, err := oauth.NewApplicationDefault(context.Background(), scope) // Application Default Credentials (ADC)
+			if err != nil {
+				return err
+			}
+
+			// parse GCP project number as header for authentication
+			var key string
+			switch c.info.uri {
+			case "trafficdirector.googleapis.com:443":
+				key = "TRAFFICDIRECTOR_GCP_PROJECT_NUMBER"
+			}
+			if projectNum := getValueByKeyFromNodeMatcher(c.nm, key); projectNum != "" {
+				c.md = metadata.Pairs("x-goog-user-project", projectNum)
+			}
+
+			c.cc, err = grpc.Dial(c.info.uri, grpc.WithTransportCredentials(creds), grpc.WithPerRPCCredentials(perRPC))
+			if err != nil {
+				return err
+			}
+			return nil
+		default:
+			return fmt.Errorf("Auto authentication mode for this platform is not supported. Please use jwt_file instead")
+		}
+	default:
+		return fmt.Errorf("Invalid authn_mode")
+	}
+}
+
+// New creates a new client
+func New() (*Client, error) {
+	c := &Client{
+		info: ParseFlags(),
+	}
+	if c.info.platform != "gcp" {
+		return c, fmt.Errorf("Can not support this platform now")
```

Comment:
Do not return client, as we were not able to create it. It must be "return nil, error" here and in other places.

Reply:
Solved.